File: leetcode/algorithms/671_second_minimum_node_in_a_binary_tree/main.java
class SecondMinimumNodeInABinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int returnSecondMinimumValue(TreeNode node, int minVal) {
        if (node == null) {
            return -1;
        }

        if (node.val > minVal) {
            return node.val;
        }

        int left = returnSecondMinimumValue(node.left, node.val);
        int right = returnSecondMinimumValue(node.right, node.val);

        if (left == -1) {
            return right;
        }
        if (right == -1) {
            return left;
        }

        return Math.min(left, right);
    }

    public int findSecondMinimumValue(TreeNode root) {
        return returnSecondMinimumValue(root, root.val);
    }

    // Solution
    public int solution(TreeNode root) {
        if (root == null) {
            return -1;
        }
        if (root.left == null && root.right == null) {
            return -1;
        }

        int left = root.left.val;
        int right = root.right.val;

        // if value same as root value, need to find the next candidate
        if (root.left.val == root.val) {
            left = findSecondMinimumValue(root.left);
        }
        if (root.right.val == root.val) {
            right = findSecondMinimumValue(root.right);
        }

        if (left != -1 && right != -1) {
            return Math.min(left, right);
        } else if (left != -1) {
            return left;
        } else {
            return right;
        }
    }
}

File: leetcode/algorithms/671_second_minimum_node_in_a_binary_tree/main.dart
import 'dart:math';

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class SecondMinimumNodeInABinaryTree {
  int returnSecondMinimumValue(TreeNode? node, int minVal) {
    if (node == null) {
      return -1;
    }

    if (node.val > minVal) {
      return node.val;
    }

    int left = returnSecondMinimumValue(node.left, node.val);
    int right = returnSecondMinimumValue(node.right, node.val);

    if (left == -1) {
      return right;
    }
    if (right == -1) {
      return left;
    }

    return min(left, right);
  }

  int findSecondMinimumValue(TreeNode? root) {
    return returnSecondMinimumValue(root, root!.val);
  }

  // Solution
  int solution(TreeNode? root) {
    List<int> res = [];

    void helper(TreeNode? node) {
      if (node == null) return;
      if (!res.contains(node.val)) {
        res.add(node.val);
      }
      helper(node.left);

      helper(node.right);
    }

    helper(root);

    if (res.length < 2) {
      return -1;
    }

    res.sort();

    return res[1];
  }
}
